language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToFunctionCodeAction.java

```diff
@@ -0,0 +1,593 @@
+import org.ballerinalang.formatter.core.Formatter;
+import org.ballerinalang.formatter.core.FormatterException;
+import org.ballerinalang.langserver.codeaction.CodeActionNodeValidator;
+import org.ballerinalang.langserver.codeaction.CodeActionUtil;
+import org.ballerinalang.langserver.codeaction.ExtractToFuncStatementAnalyzer;
+import org.ballerinalang.langserver.command.visitors.IsolatedBlockResolver;
+import org.ballerinalang.langserver.common.constants.CommandConstants;
+import org.ballerinalang.langserver.common.utils.CommonKeys;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.common.utils.FunctionGenerator;
+import org.ballerinalang.langserver.common.utils.NameUtil;
+import org.ballerinalang.langserver.common.utils.PositionUtil;
+import org.ballerinalang.langserver.common.utils.SymbolUtil;
+import org.ballerinalang.langserver.commons.CodeActionContext;
+import org.ballerinalang.langserver.commons.codeaction.spi.RangeBasedCodeActionProvider;
+import org.ballerinalang.langserver.commons.codeaction.spi.RangeBasedPositionDetails;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.CodeActionKind;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Code Action for extracting a code segment to a function.
+ *
+ * @since 2201.2.1
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class ExtractToFunctionCodeAction implements RangeBasedCodeActionProvider {
+
+    public static final String NAME = "Extract to method";
+
+    private static final String EXTRACTED_PREFIX = "extracted";
+
+    @Override
+    public boolean validate(CodeActionContext context, RangeBasedPositionDetails positionDetails) {
+        return context.currentSemanticModel().isPresent() && context.currentDocument().isPresent()
+                && CodeActionNodeValidator.validate(positionDetails.matchedCodeActionNode())
+                && (!isExpressionNode(positionDetails.matchedCodeActionNode())
+                || isExpressionExtractable(context, positionDetails));
+    }
+
+    @Override
+    public List<SyntaxKind> getSyntaxKinds() {
+        List<SyntaxKind> supportedSyntaxKinds = new ArrayList<>();
+        supportedSyntaxKinds.addAll(getSupportedStatementSyntaxKindsList());
+        supportedSyntaxKinds.addAll(getSupportedExpressionSyntaxKindsList());
+        return supportedSyntaxKinds;
+    }
+
+    @Override
+    public List<CodeAction> getCodeActions(CodeActionContext context, RangeBasedPositionDetails posDetails) {
+        if (getSupportedStatementSyntaxKindsList().contains(posDetails.matchedCodeActionNode().kind())) {
```

Comment:
This refers to supported syntaxkinds list, not the list of syntaxkinds.

Reply:
It's not a big issue. But this check is done already in the CodeActionRouter right? The requirement here is to check if the matched node is a statement and then proceed to statement logic or otherwise proceed to expression logic.

Reply:
yeah agree. will add the todo